Add documentation for using multiple `Alerts` (#1908)
hashicorp/design-system: website/docs/components/alert/partials/guidelines/guidelines.md

Use compact alerts when a less prominent alert is needed.

### Ordering Alerts by type

Alert types have an implied hierarchy and should be ordered starting from `page` and descending in hierarchy to `compact`.

> ![Ordering alerts of different types](/assets/components/alert/ordering-alerts-type.png)

Reviewer: not to be picky, but maybe this image could be something like this to convey even more the hierarchy?

<img width="1072" alt="screenshot_3499" src="https://github.com/hashicorp/design-system/assets/686239/c472da1e-e48d-4ba9-81d4-77bef8362996">

Author: Good suggestion, I've updated this image to add a bit more implied hierarchy.

## Color

<Doc::Layout @spacing="12px">
<Hds::Alert @type="inline" @color="neutral" as |A|>
<A.Title>Neutral alert title</A.Title>
<A.Description>Lorem ipsum dolar sit amet.</A.Description>
</Hds::Alert>
<Hds::Alert @type="inline" @color="highlight" as |A|>
<A.Title>Highlight alert title</A.Title>
<A.Description>Lorem ipsum dolar sit amet.</A.Description>
</Hds::Alert>
<Hds::Alert @type="inline" @color="success" as |A|>
<A.Title>Success alert title</A.Title>
<A.Description>Lorem ipsum dolar sit amet.</A.Description>
</Hds::Alert>
<Hds::Alert @type="inline" @color="warning" as |A|>
<A.Title>Warning alert title</A.Title>
<A.Description>Lorem ipsum dolar sit amet.</A.Description>
</Hds::Alert>
<Hds::Alert @type="inline" @color="critical" as |A|>
<A.Title>Critical alert title</A.Title>
<A.Description>Lorem ipsum dolar sit amet.</A.Description>

> </Hds::Alert>

Reviewer 2: Curious why it wipes out the indenting?

Author: Whoops, sometimes my Prettier config does this, I'll revert that!

- **Neutral** to provide general information to the user regarding the current context or relevant actions.
- **Highlight** to provide general or promotional information to the user prominently.
- **Success** to indicate a successful action.
  - Use the success variant sparingly. To communicate success after a user action is performed, use [Toast](/components/toast).
- **Warning** to help users avoid an issue. Provide guidance and actions, if possible.
- **Critical** to indicate critical errors that need immediate action.

### Critical alerts

Use critical alerts to convey crucial information or errors, such as validation errors that require user attention or action.

!!! Do

Use the Alert for more intrusive message communication about errors or critical disruptions at an application, page, or section level where users need to take immediate action.

<Hds::Alert @type="inline" @color="critical" as |A|>

>   <A.Title>You have exceeded 50 applies this month</A.Title>

Reviewer: @jorytindall can you revert all these reformatting changes, to reduce the noise? thanks

Author: Is there an "easy" way to do that? I could use a little assistance there, it was likely included in a commit that had other changes as a part of it.

Reviewer: I've pushed a commit that undoes these changes (only these, don't worry 😀).

However, depending on the logic within the application and the type of content being expressed to the user, it may be necessary to display more than one Alert on a page. Consider these scenarios and their accompanying guidelines if this is the case.

### Alerts of the same color

While it might not always be possible depending on the type of content and intended action, try to refrain from displaying more than one of the same `color` Alert on the same page. This adds unnecessary visual weight to the page and may result in the user being confused about which Alert is more important or which might require attention first.

![Example of multiple Alerts of the same color](/assets/components/alert/multiple-color-alerts-dont.png)

To avoid this, try combining relevant or overlapping content into a single Alert with a more concise narrative, or by translating the related messages into a list.

![Example of multiple Alerts combined into one](/assets/components/alert/multiple-color-alerts-do.png)

![Multiple Alert messages combined in a bulleted list](/assets/components/alert/multiple-color-alerts-do-bulleted-list.png)

### Alerts of the same type

> While not recommended for `page` type Alerts, it may be necessary to have more than one `inline` or `compact` Alert on a page. If this is the case, ensure that the Alerts are [ordered](#ordering-of-alerts) by impact and [contextualized](#contextualizing-alerts) in areas where they provide the most relevance.

Author: Ok I've rolled off of this term and have switched everything to "placed" or "placement"

> ### Ordering of Alerts

Author: I've moved it above the broader section for multiple alerts, but ordering really only comes into play if there _are_ multiple alerts on the page. So I'm a bit torn on that; I think the content around impact and hierarchy is more broad, though.

Order multiple Alerts by their importance and impact on the user, starting from `critical`, to communicate an Alert that must be resolved immediately, and descending in order of perceived impact to `neutral`.

![Ordering alerts](/assets/components/alert/ordering-alerts-color.png)

> Alert types have an implied hierarchy and should be ordered starting from `page` and descending in hierarchy to `compact`. This order may depend on how the Alerts are [contextualized](#contextualizing-alerts)

Author: Hmmmm, I think I must have accidentally removed or combined something here that resulted in a sentence fragment. Let me revise/rework it a bit.

![Ordering alerts of different types](/assets/components/alert/ordering-alerts-type.png)

### Contextualizing Alerts

`Inline` and `compact` Alerts can have more meaning if they are contextualized within the element that is responsible for the Alert. This can help when it’s necessary to have more than one Alert on the page and is relevant for pages that aggregate content like dashboards.

![Example of multiple Alerts contextualized to different parts of the UI](/assets/components/alert/combining-contextualized-alerts.png)

> Contextualization is helpful when displaying a specific or localized message; e.g., within a form to highlight validation errors or problems with form submission.

Author: I agree with you, and maybe it just requires a bit of rewording, but I have some content around inline alerts within forms in the [validation](https://helios.hashicorp.design/patterns/form-patterns?tab=validation#form-level-validation) section of the form patterns docs.

I could take out the specific reference to forms (make it more general)

![Placement of compact Alert](/assets/components/alert/alert-placement-compact.png =600x*)

#### Mixing Alert types

We recommend against mixing different types of Alerts in the same context, as this can unintentionally convey that different Alert types have the same hierarchical impact.

!!! Dont

Don’t mix different Alert types in the same context.

![Example of different Alert types in the same context](/assets/components/alert/mixing-alert-types-dont.png)
!!!

!!! Do

Instead, be more specific with how the Alerts are contextualized depending on what the message applies to.

![Example of different Alert types contextualized](/assets/components/alert/mixing-alert-types-do.png)
!!!

### Alerts within a Reveal

**As a last resort**, when multiple `inline` Alerts are necessary, display the most important Alert to the user within the page while obfuscating additional Alerts within a [Reveal](/components/reveal).

!!! Warning

This method should be **rare and infrequent**, and should only be used if the number of Alerts of the same type exceeds two. Instead, effort should be put into contextualizing Alerts that matter most to the user and in ways that make them easy to resolve.

> !!!

Author: I don't think I would expect multiple different Alerts in the same reveal. I'll reword the previous point to say:

> **As a last resort**, when multiple Alerts of the same **type** are necessary, display the most important Alert.

But now that I'm thinking through it, I'm not sure if I would recommend this being used for multiple compact Alerts, because those should likely be used in a more specific context. Might be unlikely that that multiple compact Alerts would be displayed next to each other.

Compact alerts can be added to a section or component or inline with content.

![Placement of compact Alert](/assets/components/alert/alert-placement-compact.png =600x*)

## Multiple alerts

In general, displaying multiple alerts on the same page is discouraged as it can negatively impact the page hierarchy and unnecessarily shift emphasis away from the main content on the page.

> However, depending on the logic within the application and the type content being expressed to the user, it may be necessary to display more than one Alert on a page. Consider these scenarios and their accompanying guidelines if this is the case.

Reviewer 3: ```suggestion
However, depending on the logic within the application and the type of content being expressed to the user, it may be necessary to display more than one Alert on a page. Consider these scenarios and their accompanying guidelines if this is the case.
```

### Alerts of the same color

!!! Dont

> Don’t display more than one of the same `color` Alert on the page. This adds unnecessary visual weight to page and may result in the user being confused about which Alert is more important or which might require attention first.

Reviewer 3: ```suggestion
Don’t display more than one of the same `color` Alert on the page. This adds unnecessary visual weight to the page and may result in the user being confused about which Alert is more important or which might require attention first.
```

![Example of multiple Alerts of the same color](/assets/components/alert/multiple-color-alerts-dont.png)
!!!

!!! Do

> Instead, try to combine relevant and overlapping content into a single Alert.

Reviewer 3: thought: I'm wondering if this statement/recommendation is fully valid only for related Alerts. If they are unrelated we could suggest separating them into distinct sections (be those paragraphs or bullets, whatever makes more sense in the context).

Author: This is a good point as well, as I'm reading through some of the feedback it seems necessary to be a bit more clear on this topic.

![Example of multiple Alerts combined into one](/assets/components/alert/multiple-color-alerts-do.png)

This may result in the length of the content being longer, but consolidating messages into a single instance of the Alert is a more concise communication pattern.
!!!

### Ordering of Alerts

> Order multiple Alerts by their importance and impact on the user, with `critical` communicating Alerts that must be resolved immediately, and descending in perceived impact to `neutral`.

Reviewer 2: ```suggestion
Order multiple Alerts by their importance and impact on the user, starting from `critical`, to communicate an Alert that must be resolved immediately, descending in order of perceived impact to `neutral`.
```

Author: That's great, thank you @KristinLBradley !

![Ordering alerts](/assets/components/alert/ordering-alerts.png)

### Alerts of the same type

> Try to avoid multiple Alerts of the same type (`page`, `inline`, etc) on the same page.

Reviewer: [blocker?] Wait. The scope of this is too wide: I clearly see cases in which I have two (or even more) alerts in the same **page**  (our documentation has a lot of pages wtih more than one alert per page :) ).

Reviewer: Another common use case is a top-level/top-of-the-page alert ("global" I think we call it) and then potentially other "local" alerts in the page.

Author: This makes me think of a couple of things:

1. This is a tough scenario to compare to our docs site, because while visually we do have multiple alerts on a page, they are actually acting more as banners. But I could see where we could get pushback on this (re: "you're using tabs for navigation, why can't we?")
2. I think I can maybe soften this language a bit and make it a bit broader. I can confidently say that there should not be more than a single page-level Alert on a page, but your point about `inline` is reasonable. I'll rework this a bit.

!!! Do

Instead, use Alerts hierarchically according to their intended prominence and importance to the user.

![An example of multiple types of Alerts on the same page](/assets/components/alert/multiple-alert-types-do.png)

> !!!

Reviewer 4: I know it wasn't the intention, but this example makes it seem like we're advocating for critical alerts to always be global. There may be instances, such as within a form, where using an inline alert for critical messages is more appropriate. Is there some way to make this more clear?

Author: I've replaced this with a `highlight` Alert to avoid this, but I'll also add some information to the contextual section that might add a bit more clarity

> ### Contextualizing Alerts

Reviewer 4: This section does feel a little duplicative of the placement section, but the content is good. There may be value in merging the two into one section.

Author: I did a bit of reorganizing here to fit this content together, though there's not really a super clean way of doing this. The upside is now relevant content is combined, but the downside is now the info on placement is a bit more buried. This might not be an issue given the naming conventions we use, but just noting it here.

Reviewer 4: Ah, I see. I had imagined that the content in "Contextualizing Alerts" would move into the Placement section (and the Placement section would remain where it was) as the placement content feels more important than multiple alerts. 

What if you dropped the heading "Contextualizing Alerts" and moved the content from that section directly into the "Placement" section? The two sections are trying to accomplish the same thing, but "placement" is a bit easier to understand as a concept. Let me know if you want to pair, happy to show you what I was thinking.

> `Inline` and `compact` Alerts can have more meaning if they are [placed](#placement) or contextualized within the element that is responsible for the Alert. This can help when it's necessary to have more than one Alert on the page and is relevant for pages that aggregate content like dashbaords.

Reviewer 3: ```suggestion
`Inline` and `compact` Alerts can have more meaning if they are [placed](#placement) or contextualized within the element that is responsible for the Alert. This can help when it's necessary to have more than one Alert on the page and is relevant for pages that aggregate content like dashboards.
```

![Example of multiple Alerts contextualized to different parts of the UI](/assets/components/alert/combining-contextualized-alerts.png)

### Alerts within a Reveal

> **As a last resort**, when multiple `inline` Alerts are necessary, display the most important Alert the user within the page, and obfuscate additional Alerts within a [Reveal](/components/reveal).

Reviewer 3: This read a bit odd, not sure my suggestions improves it though 😄 

```suggestion
**As a last resort**, when multiple `inline` Alerts are necessary, display the most important Alert to the user within the page, and obfuscate additional Alerts within a [Reveal](/components/reveal).
```

Reviewer 2: **As a last resort**, when multiple `inline` Alerts are necessary, display the most important Alert to the user within the page **_while obfuscating_** additional Alerts within a [Reveal](/components/reveal).

Reviewer 2: highlighted and bolded slightly different phrasing

Author: Thats a nice cleanup, thank you both!

Reviewer 4: +1 to @KristinLBradley's suggestion. Could probably also drop the "to the user" part to make it a bit more concise.

> **As a last resort**, when multiple `inline` Alerts are necessary, display the most important Alert the user within the page, and obfuscate additional Alerts within a [Reveal](/components/reveal).

Reviewer: I remember when we implemented the Alert in HDS, there were "stacked alerts" in Structure (imagine like stacked cards). Should we consider this as a (future) super-component to solve this pattern?

Author: Yeah maybe, though it seems like it's pretty easy to build just using our components. I guess the tricky part is probably more around the logic for displaying them in this way, thinking:

- if alerts.length =< 1 then render the alert directly in the page
- if alerts.length > 1 then render alert[0] in the page and everything else in a reveal

Could get a little tricky, but might be helpful to have this as a HDS+ pattern 😉

Reviewer 4: [This](https://www.figma.com/file/B7yd8loYS7nA2TiQcJ8UtV/Structure-1.9-%F0%9F%9A%A8-FULLY-DEPRECATED-%F0%9F%9A%A8?type=design&node-id=12569-15900&mode=design&t=lSbTh6sVKZ6zZzgl-4) is the only thing I was able to find in Structure, which appears to be the same as what Jory is proposing. @didoo I imagine what you're suggesting is something like this:

<img width="1068" alt="collapsed-alerts" src="https://github.com/hashicorp/design-system/assets/8553306/99ea927a-7387-4491-9fdf-5d8d12001237">

I'm curious about this direction because it gives the user more context on the severity of the alerts below which feels more informative and may help the user avoid missing necessary notifications.

Reviewer: @heatherlarsen unfortunately I can't get Storybook run on an old version of Structure to compare with that design, but I think it was something similar (can't remember what happened in terms of UI transition, when the user clicked on the "additional alerts").

> **As a last resort**, when multiple `inline` Alerts are necessary, display the most important Alert the user within the page, and obfuscate additional Alerts within a [Reveal](/components/reveal).

Reviewer 2: ```suggestion
**As a last resort**, when multiple `inline` Alerts are necessary, display the most important Alert the user within the page and obfuscate additional Alerts within a [Reveal](/components/reveal).
```

While it might not always be possible depending on the type of content and intended action, try to refrain from displaying more than one of the same `color` Alert on the same page. This adds unnecessary visual weight to the page and may result in the user being confused about which Alert is more important or which might require attention first.

![Example of multiple Alerts of the same color](/assets/components/alert/multiple-color-alerts-dont.png)

To avoid this, try combining relevant or overlapping content into a single Alert with a more concise narrative, or by translating the related messages into a list.

![Example of multiple Alerts combined into one](/assets/components/alert/multiple-color-alerts-do.png)

![Multiple Alert messages combined in a bulleted list](/assets/components/alert/multiple-color-alerts-do-bulleted-list.png)

### Alerts of the same type

While not recommended for `page` type Alerts, it may be necessary to have more than one `inline` or `compact` Alert on a page. If this is the case, ensure that the Alerts are [ordered](#ordering-of-alerts) by impact and [placed](#placing-alerts) in areas where they provide the most relevance.

### Alerts within a Reveal

> **As a last resort**, when multiple Alerts of the same type are necessary, display the Alerts that require immediate attention or have the most impact to the user within the page while obfuscating additional Alerts within a [Reveal](/components/reveal). We recommend against hiding `critical` and `warning` Alerts in this manner as they generally must be addressed quickly to avoid a negative outcome.

Reviewer 4: I'm still unsure about this as a pattern recommendation. What do you think about adding a note and asking teams to include this in future product research and share the findings with us? Or some other way to help elicit more feedback around this pattern.

Author: I'm not opposed to removing it, I can add a note in a banner or something, but this it does seem helpful to have something like this as a "catch-all" or just-in-case.

If we're unsure about it (and think that it's unlikely to  happen frequently, if at all), then it might just be better to remove it entirely so that the idea doesn't exist out in the wild. What do you think?

Reviewer 4: I think it's fine to keep it (this was ultimately the goal of the original ticket, after all), but I do think it's worth noting that we'd like real user feedback on this pattern from teams that do end up needing to use it.

Author: Added a bit of a note here requesting that consumers perform usability testing and share their findings with the team.
